modules/p2/ding_interests_by_history/ding_interests_by_history.module

```diff
@@ -0,0 +1,304 @@
+
+  // Get tags vocabulary.
+  $tags_vocabulary = taxonomy_vocabulary_machine_name_load('ding_content_tags');
+  // Get selected terms.
+  $node_query = new EntityFieldQuery();
+  $node_query->entityCondition('entity_type', 'taxonomy_term')->propertyCondition('vid', $tags_vocabulary->vid)->fieldCondition('field_status', 'tid', $selected_term_id[0]);
+  $result = $node_query->execute();
+
+  $terms = array();
+  if ($result) {
+    $terms = array_keys($result['taxonomy_term']);
+  }
+  // Get users interests.
+  $list = ding_list_get_list(DING_INTERESTS_LIST);
+  $elements = ding_list_load_elements($list);
+  $interests = array();
+  foreach ($elements as $element) {
+    $term = taxonomy_get_term_by_name($element->field_search_title['und'][0]['value']);
+    $tid = key($term);
+    $interests[] = $tid;
+  }
+
+  // Remove the terms already selected by the user from the available terms.
+  $terms = array_diff($terms, $interests);
+
+  // Load terms to get their names for a comparison with the subjects,
+  // genres and types of the ting objects.
+  $terms = taxonomy_term_load_multiple($terms);
+  // List terms by tid.
+  $term_names = array();
+  foreach ($terms as $tid => $term) {
+    $term_names[$tid] = strtolower($term->name);
+  }
+
+  return $term_names;
+}
+
+/**
+ * Get terms which the user might be interested in based on their loan history.
+ *
+ * @return array
+ *   The returned array contains a list of ding_content_tags,
+ *   which correspond to loans previous made by the user.
+ *   These elements are keyed by their tid and their value is the term name.
+ */
+function ding_interests_by_history_suggested_terms() {
+  if (!ding_user_consent_has_consent('loan_history_store')) {
+    return array();
+  }
+  // Get available terms. Return empty if none are recommended.
+  $term_names = ding_interests_by_history_get_selected_terms();
+  if (empty($term_names)) {
+    return array();
+  }
+
+  // Reverse the list to get the most recent first.
+  $object_ids = array_reverse(ding_interests_by_history_get_loan_entities());
+  
+  $object_ids = array_slice($object_ids, 0, DING_INTERESTS_BY_HISTORY_MAX_AMOUNT);
+  // Load ting objects to access their data.
+  $ting_objects = ding_entity_load_multiple($object_ids);
+
+  // Merge subjects, genres and types into array for comparison with tag list.
+  $historical_terms = array();
+  foreach ($ting_objects as $tojb) {
+    $subject = $tojb->getSubjects();
+    $historical_terms = array_merge($historical_terms, $subject);
+
+    // $type = $tojb->getType();
+    // $historical_terms = array_merge($historical_terms, array($type));
+    $record = $tojb->getRecord();
+    if (isset($record['dc:subject']['dkdcplus:genre'])) {
+      $genre = $record['dc:subject']['dkdcplus:genre'];
+      $historical_terms = array_merge($historical_terms, $genre);
+    }
+  }
+
+  // Get only unique elements.
+  $unique_terms = array_count_values($historical_terms);
+
+  return $unique_terms;
```

Comment:
This return value does not seem to comply with the DocBlock: 

> The returned array contains a list of ding_content_tags, which correspond to loans previous made by the user. These elements are keyed by their tid and their value is the term name.

As far as I can tell it returns an array with names as key and frequency as values.